LIN_ADVANCE with unified stepper ISR

Sebastianv650 opened the pull request:
This is a test version of a `LIN_ADVANCE` without an own estepper ISR.

I think the Timer 0 currently used by `ADVANCE `and `LIN_ADVANCE` is a possible source of some errors some user get when using this features. This PR shares the high-resolution Timer 1 with all steppers, therefore the advance steps should be distributed in time more equally, hopefully resolving some issues.

It's not well tested (2 prints finished, no problems), but up to now I'm optimistic. Some problems like rattling extruder in arcs and extruder starvation at high e-only moves seems to be gone.

Hopefully @VanessaE and others will gives this a try?

Commit:
1st draft of unified stepper ISR

thinkyhead commented:
This is a very good idea. No more dueling interrupts fighting for time.

Sebastianv650 commented:
So far no problems. Cleanup is now done.
But my printer was always playing nice with this, therefore I'm shy to say yes, merge it.

Found some other possible volunteers:
@landodragon141
@Kaibob2

landodragon141 commented:
I would be very happy to test this but I'm still waiting on my new control board after my son let the magic smoke out of my current one.

Kaibob2 commented:
![20161213_214507_1481661964881_resized](https://cloud.githubusercontent.com/assets/20832276/21158532/d2de87e0-c17d-11e6-9d61-d81f73c9119d.jpg)
![20161213_214522_1481661962404_resized](https://cloud.githubusercontent.com/assets/20832276/21158535/d544f014-c17d-11e6-8b78-9065a1196fb0.jpg)

I would say this looks great for the first print.
Printed at:
70mm/s everything
A = 500 mm/s²
XYJerk = 10 mm/s
Bowden 500mm

Bottom K=200
Above K=175
Above K=150
Top K=100

Acceleration / Speed is too high, have some ringing, but no rattling Extruder or anything suspicious. Everything runs really smooth.

landodragon141 commented:
That looks really good. I'm curious what @VanessaE results will be, based on what I recall she has had the most dramatic issues.

VanessaE commented:
I can't do any real testing for a while.  A lot of the issues I have with my machine, I am certain, have to do with junk hotend parts, so I've sprung for a genuine J-head.  I'll wait for that to arrive and get to printing with it before I fiddle with LIN_ADVANCE.

Best not to wait for me, if this code works as intended, you may as well merge it for wider testing.  I'll report back later if there's still a problem.

thinkyhead commented on Marlin/stepper.cpp:
```diff
@@ -328,13 +329,42 @@ void Stepper::set_directions() {
-ISR(TIMER1_COMPA_vect) { Stepper::isr(); }
+ISR(TIMER1_COMPA_vect) {
+  #if ENABLED(ADVANCE) || ENABLED(LIN_ADVANCE)
+    Stepper::ISRhandler();
```
I think `Stepper::ISRhandler` is short enough that you can just put its code here and drop the extra method call.

Sebastianv650 replied:
Oh, I just deleted my answer by accident..
I tried this first, but had problems with it. First I got the message `nextMainISRrun` and others are not in scope. After changing to `Stepper::`, it's complaining the variables are private. Do you have an idea how to solve this in a clean way?

thinkyhead replied:
Not super clean. If the variables needed are out of scope, you can precede them with `stepper.` and if they are `private`, you can make them `protected` instead.

For now let's just leave things as they are.

thinkyhead commented on Marlin/stepper.cpp:
```diff
@@ -346,11 +376,13 @@ void Stepper::isr() {
+      nextMainISRrun = 200; // Run at max speed - 10 KHz
+    #else
+      OCR1A = 200; // Run at max speed - 10 KHz
     #endif
```
Here's a good place for a macro (or two) to simplify the code…
```cpp
#if ENABLED(ADVANCE) || ENABLED(LIN_ADVANCE)
  #define _NEXT_ISR(T) nextMainISRrun = T
#else
  #define _NEXT_ISR(T) OCR1A = T
#endif
#define _NEXT_ISR_GO(T) _NEXT_ISR(T); cli()
```
```cpp
_NEXT_ISR_GO(200);
...
_NEXT_ISR_GO(2000);
...
_NEXT_ISR_GO(2000);
...
_NEXT_ISR(timer);
...
_NEXT_ISR(OCR1A_nominal);
```

thinkyhead commented on Marlin/stepper.cpp:
```diff
@@ -688,7 +737,12 @@ void Stepper::isr() {
+      }
+      else {
+        eISR_Rate = 65535; // don't execute the advance ISR
+      }
```
Also fine… (i.e., following the "least-repetition" guideline).
```cpp
e_ISR_Rate = !e_steps[TOOL_E_INDEX] ? 65535 : timer * step_loops / abs(e_steps[TOOL_E_INDEX]);
```

thinkyhead commented on Marlin/stepper.cpp:
```diff
@@ -698,40 +752,46 @@ void Stepper::isr() {
+      }
+      else {
+        eISR_Rate = 65535; // don't execute the advance ISR
+      }
```
Et voila.
```cpp
e_ISR_Rate = !e_steps[TOOL_E_INDEX]
             ? 65535
             : OCR1A_nominal * step_loops_nominal / abs(e_steps[TOOL_E_INDEX]);
```

thinkyhead commented on Marlin/stepper.h:
```diff
@@ -105,8 +105,9 @@ class Stepper {
-      static volatile uint8_t eISR_Rate;
+      static volatile uint16_t nextMainISRrun;
+      static volatile uint16_t nextAdvanceISRrun;
+      static volatile uint16_t eISR_Rate;
```
As long as these variables aren't accessed outside the ISR (and inside `CRITICAL_SECTION`) , they don't need to be declared '`volatile`' here. Removing `volatile` will save a few bytes in compilation because they're read more than once in `ISRhandler` itself. To ensure that they're safe for the single task, put them in the `private` data section.

Kaibob2 commented:
![1](https://cloud.githubusercontent.com/assets/20832276/21174493/c7e7641e-c1dd-11e6-8017-1b35ab19e96d.jpg)
![2](https://cloud.githubusercontent.com/assets/20832276/21174494/c7ea246a-c1dd-11e6-9b4a-f812773737e9.jpg)
![3](https://cloud.githubusercontent.com/assets/20832276/21174495/c805abea-c1dd-11e6-81fb-c4600ca31a8d.jpg)

0,2mm layer height
Retract 1mm@35mm/s
100mm/s everything !!
A = 1000 mm/s²
XYJerk = 10 mm/s
Bowden 500mm

Especially the gecko came out great. No stringing or any of the problems reported some time ago [here](https://github.com/MarlinFirmware/Marlin/issues/5222#issuecomment-261186913)

The gecko is a little bit shaky because of the high speed, but in this particular object it even looks like good :)

So, from my point of view this is a huge improvement and can be merged.

Kaibob2 commented:
@Sebastianv650 Maybe you should consider modifying (or deleting) the formula to calculate K as for my setup for example a K of 235 would be suitable according to it. 
In reality with a K>150 i suffer from massive underextrusion where a loop starts. K>200 is unusable.

Sebastianv650 commented:
@Kaibob2 
Thanks for testing, I'm more than happy that it's working for you! Looks like I did it right this time, finaly ;-)
I already wanted to delete the formula before, now it's good time to do it. I will replace it by a note to report calibrated K values with some printer data - I think if I get some more responses it should be possible to get a working formula again.

@thinkyhead 
I will go through your comments now.

fiveangle commented:
Working but shows weird artifacts from what looks like not relieving enough pressure on layer change, creating blobs. Need to go back and test same gcode on RC8

fiveangle commented:
Stock Printrbot Simple Metal 1403.  ~3 week old Pre-RC8 RCBugFix on left

![img_3407](https://cloud.githubusercontent.com/assets/2219447/21196004/f832095a-c1ea-11e6-8aed-811fe59f3ff2.JPG)
![img_3406](https://cloud.githubusercontent.com/assets/2219447/21196005/fae8ee66-c1ea-11e6-8609-656cdeeb0b41.JPG)

spd/acc/j 70/1000/5 retraction 1.2mm@40 (on poorer one in photos)

I'll up retraction to 2mm@60 (which may be what the "K75" one from earlier was printed at, but I thought it was 1mm@60), spd to 80 then retest on this branch and then RC8 with same gcode.

Sebastianv650 commented:
In my print today I can also see 1-2 possible errors that were not present without this change.
I have to do a little bit of investigation.

Sebastianv650 commented:
@fiveangle the left one in the pictures is with current advance code and right one with this pr?

fiveangle commented:
@Sebastianv650 - yes, in pics above, one on left is Marlin RCBugFix from 14 days ago and right is this PR.  I found the same GCODE used to print the left model in my Octoprint server so just reprinted it using your PR.  It looks pretty much the same so I think the above test was non-optimal slicer settings.

The pics below are same GCODE on same printer Marlin RCBugFix from 14 days ago (left) and this PR (right) and the only thing that jumps out is the weird artifact to the lower-right of one of the concave sphere:

![img_3409](https://cloud.githubusercontent.com/assets/2219447/21202206/c625a3a0-c202-11e6-959c-aed995b8e15f.JPG)
![img_3408](https://cloud.githubusercontent.com/assets/2219447/21202212/cd4d65b4-c202-11e6-9d91-64a7c456e78b.JPG)

I don't see how that could be blobs or anything.  I'm reprinting it now...

fiveangle commented:
Must have been something messed up on the infill, but reprint looks good and essentially indistinguishable from current LINEAR_ADV code #shipit ;)

Sebastianv650 commented:
I don't like the layers in your print where it looks like a part of the perimeter is missing. I got the same thing in my print yesterday. I was printing a lot units with this gcode before without this issue. Even worse, Only One of two prints have this?

Sebastianv650 commented:
Touched the wrong point on my smartphone,don't wanted to close this.
As said, only one of my two prints from yesterday has this under extrusion. I know this behavior from early experiments where i had issues with LIN_ADVANCE and retraction, for example this can be when their are missing steps during unretract or if a blob is somewhere in the print as on your picture. 
Can be caused by too aggressive retract settings, but i don't changed something.

Up to now i can't reproduce this safely. If you find a peace of gcode that shows this issue always, please tell me.

Sebastianv650 commented:
@fiveangle do you use FW retraction? And did you print the test parts over USB, with or without communication errors?

I checked for possible overflows and a few other things, but found nothing up to now. Completely random errors are not nice to find..

Commit:
Remove formula from description

Sebastianv650 commented:
@fiveangle Oh I think I mixed up your old/new part order again. The one with the partly "missing perimeter" is without this PR? If true, your prints are clean now.

Maybe I'm hunting a serial communication ghost, changed the UART ISR workaround places.

VanessaE commented:
I should point out, it looks like @fiveangle used Slic3r.  If so, you're seeing a bug in how Slic3r places seams, making those "ripples" look worse than they otherwise would be.  That can easily look like random glitches in Marlin.

fiveangle commented:
@Sebastianv650 I'm not sure what you mean by "missing perimeter", but on this printer I use Octoprint over USB at 115k w/o FW retract.  In ALL pictures, print on LEFT is RCBugFix from 2 weeks ago; print on **RIGHT is this PR**.  I was under the impression the blobs on left prints may be due to over-taxed Teensy2.0++ built into Printrboard when LINEAR_ADV is enable causing the print to pause occasionally when serial transfer takes too many resources (I didn't babysit, so don't know if the blobs were due to that).  I later moved from 250k to 115k comms speed to try and help if that were the case.  SD card support is enabled but not used, LCD disabled (as PBSM doesn't have one, and Teensy can't enable LINEAR BL, LINEAR_ADV, SD, and LCD at the same time anyway due to resource constraints).

When I refreshed Octoprint page in a browser where OP had gone to "sleep" mode yesterday during this 100mm/s print (https://youtu.be/dkVFFfI7aMY) I got studdering in the print until the browser refresh completed, but I see that as really an Octoprint issue.  Setting the "no sleep" option in OP would likely resolve as I think it drops serial comms during "sleep" so the re-establish of comms likely caused the studdering (?).  I plan to jump to a RPi 3 (running RPi v2 now) and a faster SD card for it to see if that helps.

@VanessaE print on LEFT (14-day-old RCBugFix LINEAR_ADV) was sliced with Simply3D 3.1.0 and print on RIGHT (this PR) sliced with S3D 3.1.1.

I'm working on printing a bunch of trinket gifts for an eggnog party at my house tomorrow and printing as fast as possible, so will let you know if I run into any obvious problems.

Thanks,

-=dave

Sebastianv650 commented:
![unbenannt](https://cloud.githubusercontent.com/assets/10776877/21238904/13e353a0-c305-11e6-9737-224d10dea0e8.JPG)

That's what I mean with partialy missing perimeter.

OK I will risk some meter of filament to see if I get print errors in the next 2 days. Let's hope the best.

fiveangle commented:
I will specifically look for those issues on the prints I'm running with this PR.  So far, I've not seen any.

Commit:
Fix stepper.h

Do we need this lines at all? OCR1A will be reset in the stepper ISR in
any circumstates..

Sebastianv650 commented:
A lot of checks and test prints later, I would say it's working at least as good as the current implementation. I'm very sure the perimeter error I got was due to tangled filament on the spool. I got no further one up to now.
I would say merge it and see how it behaves on other machines.

Blue-Marlin commented:
> I would say merge it and see how it behaves on other machines.

I'd say - the ideal order is the other way around. 
But i know - its always had to find qualified testers for a PR.

Sebastianv650 commented:
@Blue-Marlin true, but I have the feeling the amount of users of this feature is around 5, and not everybody is looking into the PR section. I tried to mention everybody I know here, and so far every one except me seems to have no problem - and mine was a filament feed error.
So we can eighter keep this PR open and wait if somebody else will report while I continue to use it for my prints, watching the result. I have no problem with that at all. Or we merge it, risking a undo of this PR as a worst-case.

If you prefer to wait some time, I can say I will have quite a lot of prints todo until Friday, then I will be away for 1-2 weeks. That means, if something will happen from my side, it's within the next 4 days.

Kaibob2 commented:
I would also vote for merge, as i did some test here and there with the recent PR here and had no critical issues. But, like always the "printer modify list" is long and lin_advance testing is, sorry @Sebastianv650, not on top :)

fiveangle commented:
> I'm very sure the perimeter error I got was due to tangled filament on the spool.

I find that very hard to believe, as these "perimeter error" only surfaced on my print immediately after enabling LINEAR_ADV in RC7 from 3 weeks ago and absolutely did not have any filament feeding issues.  Like you, I wish I knew why they were so prevalent.  If your theory of this being emergent from some retract settings when used with LINEAR_ADV, they will eventually surface again and we deal with it then since whatever the cause, it's absolutely not due to this PR since it was present before.

That said, I too vote for opening up to more people.  I seem to get a few more communication errors recently while connecting to a Printrboard from Octoprint, even at 115k, but I suspect these are red herring related to recent OP 1.3 release which was updated near the same exact time.  Print-wise, with approx 30hrs of printing, this PR is working identically to RC8 in my environment, if not better (as "perimeter error" issues  with this PR have been non-existent).

Caveat: I've not tried with the new conditional omitting set of OCR1A when pressure control enabled.

Sebastianv650 commented:
@fiveangle 

> I find that very hard to believe, as these "perimeter error" only surfaced on my print immediately after enabling LINEAR_ADV in RC7 from 3 weeks ago and absolutely did not have any filament feeding issues.

Yes, but in the print with this PR the problem wasn't there I thought? It should not have a relation to retracts, but problems might occur with retracts in the current implementation as it's using another timer for the ISR. It's theoreticaly possible that the timer isn't firing in equal distances because the other timer used for the main ISR has a higher priority. Second, this second timer has a lower resolution, therefore it might be not as good for precise stepper pulses as the main timer.

thinkyhead commented on Marlin/example_configurations/TAZ4/Configuration_adv.h:
```diff
@@ -553,14 +553,7 @@
- * ----------------------------|-----------|------------|
- * Stiff filament (PLA)        | K=47*L/10 | K=139*L/10 |
- * Softer filament (ABS, nGen) | K=88*L/10 | K=260*L/10 |
+ * See Marlin documentation for calibration instructions.
```
Shall we post this guide on marlinfw.org? 😄  Starting to fill the site, slowly…

Sebastianv650 replied:
Not this one, as it's not valid anymore. If this feature is used by more people in the future and we can get some feedback about used filament (diameter, material), estep/mm setting and calibrated K values we might find a better one.

thinkyhead commented:
Will merge shortly as #5573

VanessaE commented:
I've been out of town for a few days and finally got a chance to give this a try (by way of a clean clone + transcribing my settings).  Sorry to say that even at K=75, the extruder rattles.  A lot.

I should mention, btw, that I am not using the minimum stepper pulse feature anymore. That said, setting it to 4 like I used to doesn't stop the rattling.

Also, on a good note, I'm not seeing any transmission errors anymore.

Sebastianv650 commented:
Is it only about the noise, or is it also skipping steps and if yes it's worse than before?

I'm also not at home since more than a week, but I will configure and print with latest RCBugFix in 2-4 days again.

VanessaE commented:
I didn't let it run long enough to check for lost steps and such, however with that much noise, well...  where there's noise, there's wear and tear.

landodragon141 commented:
@VanessaE You use a geared extruder right? Is it possible it's just the slop in the gears?

VanessaE commented:
It is a geared extruder, and while there is a *little* slop in the gears, it's not *that* bad.  Honestly it sounds a lot like lack of acceleration.

VanessaE commented:
@Sebastianv650 what became of that last set of tests you mentioned?  I'm really curious why it rattles so much for me (as of 922c67f17d5052f499f331f9702559799c6a1b0c anyway), or rather, what's the right way to solve it.
